src/profile-logic/profile-data.js

```diff
@@ -197,12 +193,32 @@ export function getCallNodeInfo(
           ? frameTable.nativeSymbol[frameIndex]
           : null;
       const funcIndex = frameTable.func[frameIndex];
-      const prefixCallNodeAndFuncIndex = prefixCallNode * funcCount + funcIndex;
 
       // Check if the call node for this stack already exists.
-      let callNodeIndex = prefixCallNodeAndFuncToCallNodeMap.get(
-        prefixCallNodeAndFuncIndex
-      );
+      let callNodeIndex;
+      const currentLastSibling =
+        prefixCallNode === -1
+          ? currentLastRoot
+          : currentLastChild[prefixCallNode];
+      if (currentLastSibling !== -1) {
+        if (func[currentLastSibling] === funcIndex) {
+          callNodeIndex = currentLastSibling;
+        } else {
+          const currentFirstSibling =
+            prefixCallNode === -1 ? 0 : firstChild[prefixCallNode];
+          for (
+            let currentSibling = currentFirstSibling;
+            currentSibling !== currentLastSibling;
+            currentSibling = nextSibling[currentSibling]
+          ) {
+            if (func[currentSibling] === funcIndex) {
+              callNodeIndex = currentSibling;
+              break;
+            }
+          }
+        }
+      }
```

Comment:
Could we simplify this with just:
```js
          const currentFirstSibling =
            prefixCallNode === -1 ? 0 : firstChild[prefixCallNode];
          for (
            let currentSibling = currentFirstSibling;
            currentSibling !== -1;
            currentSibling = nextSibling[currentSibling]
          ) {
            if (func[currentSibling] === funcIndex) {
              callNodeIndex = currentSibling;
              break;
            }
          }
```
I think the first `if` block with `currentLastSibling` is included in the `for` loop if we loop until `currentSibling` is `-1`.
But maybe you wanted to separate this case for performance reason? Is it common enough that the last child is the right function that it's worth singularizing this case? In that case please state it in a comment.

Reply:
I'm having a hard time simplifying it without making it more complicated. The issue is about what to do for the very first call node. `firstChild[-1]` is not something we want to access, and `nextChild[0]` also hasn't been set yet. Here's what I ended up with (ignoring the lint warning about nested ternaries):

```js
      const currentFirstSibling =
        prefixCallNode === -1
          ? stackIndex === 0
            ? -1
            : 0
          : firstChild[prefixCallNode];
      for (
        let currentSibling = currentFirstSibling;
        currentSibling !== -1;
        currentSibling = nextSibling[currentSibling]
      ) {
        if (func[currentSibling] === funcIndex) {
          callNodeIndex = currentSibling;
          break;
        }
      }
```

Another option would be to set `nextSibling[0] = -1;` before entering the loop, but that's awkward, too. I think I'll just add a comment that we check the last sibling first and call it a day.

Reply:
I'm going with this:

```js
      if (stackIndex !== 0) {
        const currentFirstSibling =
          prefixCallNode === -1 ? 0 : firstChild[prefixCallNode];
        for (
          let currentSibling = currentFirstSibling;
          currentSibling !== -1;
          currentSibling = nextSibling[currentSibling]
        ) {
          if (func[currentSibling] === funcIndex) {
            callNodeIndex = currentSibling;
            break;
          }
        }
      }
```